### PR TITLE
Generic Config Updater Syslog server test

### PR DESCRIPTION
**Pull request by wen587**
### Description of PR

Summary: Testcase of syslog server for generic updater apply-patch
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
[End to End test support for Generic Updater apply-patch](https://github.com/ghooo/SONiC/blob/c1f3f3b5427d0cafb3defd93df8b906a26fcee8a/doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md#91-unit-tests-for-apply-patch)
This PR is to verify the usage of 'config apply-patch' works on syslog server
#### How did you do it?
First we setup clean syslog env. Then make some config apply change. And check if the syslog server is changed as expected.
Updated start-limit-hit function in `gu_utils.py` 

#### How did you verify/test it?
Run test of sonic-mgmt/tests/generic_config_updater/test_syslog.py on KVM

### CODE REVIEW

**Comment by wen587** [comment]
[sonic-syslog.yang](https://github.com/Azure/sonic-buildimage/pull/9265) is needed to complete the test.

File: tests/generic_config_updater/test_syslog.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
+import logging
+import pytest
+
+from tests.common.helpers.assertions import pytest_assert
+from tests.common.config_reload import config_reload
+from tests.generic_config_updater.gu_utils import apply_patch, expect_op_success_and_reset_check, expect_res_success, expect_op_failure
+from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+
+pytestmark = [
+    pytest.mark.topology('t0'),
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:10]: >t0 [](http://example.com/codeflow?start=26&length=2)

syslog is a feature for any topology, not just `t0`.

File: tests/generic_config_updater/gu_utils.py [hunk 2]
```diff
@@ -71,29 +73,32 @@ def start_limit_hit(duthost, container_name):
...
     pytest_assert(
         not service_start['rc'],
         "{} systemctl start service fails"
     )
 
-    reset_container = wait_until(threshold,
+    if not service_name in CONTAINER_SERVICES_LIST:
+        return
+
+    reset_service = wait_until(threshold,
```
Comment (qiluo-msft) [inline tests/generic_config_updater/gu_utils.py:96]: >threshold [](http://example.com/codeflow?start=31&length=9)

change all the names: threshold->timeout

File: tests/generic_config_updater/test_syslog.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+]
+
+logger = logging.getLogger(__name__)
+
+# This is restricted by sonic-syslog.yang
+SYSLOG_MAX_SERVER=-1
+
+SYSLOG_THRESHOLD=10
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:18]: >10 [](http://example.com/codeflow?start=17&length=2)

Add document about its unit, is it in seconds?

File: tests/generic_config_updater/test_syslog.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    Syslog Servers
+    ----------------
+    [10.0.0.1]
+    ...
+    [10.0.0.SYSLOG_MAX_SERVER]
+    """
+    if SYSLOG_MAX_SERVER == -1:
+        pytest.skip("SYSLOG_MAX_SERVER is not set")
+
+    syslog_servers = ["10.0.0.{}".format(i) for i in range(1, SYSLOG_MAX_SERVER+1)]
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:217]: >SYSLOG_MAX_SERVER [](http://example.com/codeflow?start=62&length=17)

What if SYSLOG_MAX_SERVER >= 256? invalid IP?

File: tests/generic_config_updater/test_syslog.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+
+pytestmark = [
+    pytest.mark.topology('t0'),
+]
+
+logger = logging.getLogger(__name__)
+
+# This is restricted by sonic-syslog.yang
+SYSLOG_MAX_SERVER=-1
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:16]: >-1 [](http://example.com/codeflow?start=18&length=2)

Explain -1 in code comment?

File: tests/generic_config_updater/test_syslog.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+SYSLOG_THRESHOLD=10
+SYSLOG_INTERVAL=1
+
+@pytest.fixture(scope="module")
+def cfg_facts(duthosts, rand_one_dut_hostname):
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:22]: >cfg_facts [](http://example.com/codeflow?start=4&length=9)

Is it possible to reuse this function? Seem duplicated in the code base.

File: tests/generic_config_updater/test_syslog.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    Args:
+        duthosts: list of DUTs.
+        rand_selected_dut: The fixture returns a randomly selected DuT.
+        cfg_facts: config facts for selected DUT
+    """
+    duthost = duthosts[rand_one_dut_hostname]
+
+    config_tmpfile = generate_tmpfile(duthost)
+    logger.info("config_tmpfile {}".format(config_tmpfile))
+    logger.info("Backing up config_db.json")
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:39]: combine 2 info lines into one?

File: tests/generic_config_updater/test_syslog.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    duthost.shell("sudo cp /etc/sonic/config_db.json {}".format(config_tmpfile))
+
+    # Cleanup syslog server config
+    syslog_servers = cfg_facts.get('SYSLOG_SERVER', {})
+    for syslog_server in syslog_servers:
+        duthost.shell("sudo config syslog del {}".format(syslog_server),
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:45]: >shell [](http://example.com/codeflow?start=16&length=5)

Check the return code?

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:18]: >= [](http://example.com/codeflow?start=16&length=1)

Use consistent coding style for assignment statement.

File: tests/generic_config_updater/test_syslog.py (hunk 1 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py]: change to 'any'

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py:15]: I am not sure about this. From my test, it doesn't take much time to do that. So I just give it a short threshold.

File: tests/generic_config_updater/test_syslog.py (hunk 4 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py:212]: Where does 256 come from? It seems we dont have a max from sonic-syslog.yang.
Invalid IP was supported in tc4.

File: tests/generic_config_updater/test_syslog.py (hunk 5 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py]: Done

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py:15]: I am not sure what does it mean. Eliminate the empty line?

File: tests/generic_config_updater/test_syslog.py (hunk 8 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py]: Added the check. But I am confused that when do we need to check the return code? Seems not every command is being checked.

File: tests/generic_config_updater/test_syslog.py (hunk 4 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py:215]: Got your point. I set a test max upper limit. 
I will change related code if the syslog server maximum in yang is over that limit.

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (qiluo-msft) [reply tests/generic_config_updater/test_syslog.py:15]: Sorry, I mean the blanks before and after `=`

Comment (qiluo-msft) [reply tests/generic_config_updater/test_syslog.py:15]: Sorry I mean the unit of 10? Is it seconds, not milliseconds or others?

File: tests/generic_config_updater/test_syslog.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+from tests.generic_config_updater.gu_utils import apply_patch, expect_op_success_and_reset_check, expect_res_success, expect_op_failure
+from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+
+pytestmark = [
+    pytest.mark.topology('any'),
+]
+
+logger = logging.getLogger(__name__)
+
+SYSLOG_THRESHOLD=10
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:15]: >THRESHOLD [](http://example.com/codeflow?start=7&length=9)

Name it as *_TIMEOUT?

File: tests/generic_config_updater/test_syslog.py (hunk 8 above)
Comment (qiluo-msft) [reply tests/generic_config_updater/test_syslog.py]: Ideally we need to check every return code in case of any disaster. In some special case, if you don't care about any disaster/corner cases and just want to continue, then it's okay to ignore the return code.

For example, in this case, `config` command will possible return an error code if the system is so busy. You should treat it as a test case failure.

**Review by qiluo-msft — COMMENTED** [review]
Minor issues remain

File: tests/generic_config_updater/test_syslog.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    for syslog_server in syslog_servers:
+        del_syslog_server = duthost.shell("sudo config syslog del {}".format(syslog_server),
+            module_ignore_errors=True)
+        pytest_assert(not del_syslog_server['rc'],
+            "syslog server '{}' is not deleted successfully".format(syslog_server))
+
+    yield
+
+    logger.info("Restoring config_db.json")
+    duthost.shell("sudo cp {} /etc/sonic/config_db.json".format(config_tmpfile))
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_syslog.py:49]: >cp [](http://example.com/codeflow?start=24&length=2)

We also need to restore the Redis database in memory with command line `sudo config reload -y`.

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py:15]: Its unit is second

File: tests/generic_config_updater/test_syslog.py (hunk 9 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py]: Changed

File: tests/generic_config_updater/test_syslog.py (hunk 3 above)
Comment (wen587) [reply tests/generic_config_updater/test_syslog.py]: Added.

**Review by qiluo-msft — APPROVED** [review]
